test/test_sum_product.py

```diff
@@ -110,58 +110,87 @@ def test_partial_sum_product(impl, sum_op, prod_op, inputs, plates, vars1, vars2
+                    markov_to_step[k] |= frozenset(
+                        ('{}_{}_{}'.format(var, markov_plate, i+1)
+                         for i in range(duration))
                     )
-                    reduce_vars |= frozenset(
-                        ('{}_{}'.format(curr_to_drop[var], i+1) for
-                         var in curr_to_drop.keys() for i in range(factor.inputs[markov_plate].size))
-                    )
-                    reduce_vars |= frozenset(
-                        ('{}_{}'.format(prev_to_drop[var], i) for
-                         var in prev_to_drop.keys() for i in range(factor.inputs[markov_plate].size))
-                    )
-                    slice_factors = [factor(
-                        **{markov_plate: i},
-                        **{var: '{}_{}'.format(var, i+1) for var in local_vars},
-                        **{var: '{}_{}'.format(curr_to_drop[var], i+1) for var in curr_to_drop.keys()},
-                        **{var: '{}_{}'.format(prev_to_drop[var], i) for var in prev_to_drop.keys()}
-                        ) for i in range(factor.inputs[markov_plate].size)]
-                    unrolled_factors.extend(slice_factors)
-        else:
-            unrolled_factors.append(factor)
-
-    expected = sum_product(sum_op, prod_op, unrolled_factors, reduce_vars, plates)
+        reduce_vars -= frozenset(
+            (var for var in prev_to_drop.keys())
+        )
+        reduce_vars -= frozenset(
+            (var for var in curr_to_drop.keys())
+        )
+        for i in range(duration):
+            reduce_vars |= frozenset(
+                ('{}_{}_{}'.format(var, markov_plate, i+1) for
+                 var in local_vars)
+            )
+            reduce_vars |= frozenset(
+                ('{}_{}_{}_curr'.format(var, markov_plate, i+1) for
+                 var in local_markov_vars)
+            )
+            reduce_vars |= frozenset(
+                ('{}_{}_{}_prev'.format(var, markov_plate, i+1) for
+                 var in local_markov_vars)
+            )
+            reduce_vars |= frozenset(
+                ('{}_{}'.format(curr_to_drop[var], i+1) for
+                 var in curr_to_drop.keys())
+            )
+            reduce_vars |= frozenset(
+                ('{}_{}'.format(prev_to_drop[var], i) for
+                 var in prev_to_drop.keys())
+            )
+        for factor in factors:
+            if markov_plate in factor.inputs:
+                slice_factors = [factor(
+                    **{markov_plate: i},
+                    **{var: '{}_{}_{}'.format(var, markov_plate, i+1)
+                       for var in local_vars},
+                    **{'{}_curr'.format(var): '{}_{}_{}_curr'.format(var, markov_plate, i+1)
+                       for var in local_markov_vars},
+                    **{'{}_prev'.format(var): '{}_{}_{}_prev'.format(var, markov_plate, i+1)
+                       for var in local_markov_vars},
```

Comment:
To clarify, is the purpose of these two new substitutions to handle cases like the current, tractable version of `test_modified_partial_sum_product_14` where one time variable is completely nested within a single step of another?

Reply:
Yep, that's right.